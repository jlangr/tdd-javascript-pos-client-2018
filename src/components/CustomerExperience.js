import React, { Component } from 'react';
import { Form, FormGroup, FormControl, ControlLabel, Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import Table from 'rc-table'; // see https://github.com/react-component/table 

import { createCheckout, postItem } from '../actions';

export class CustomerExperience extends Component {
  constructor() {
    super();
    this.itemTableColumns = this.createTableColumns();
    this.state = { scanCode: '' };
  }

  createTableColumns() {
    return [
      { title: 'Description', dataIndex: 'description', key: 'description', width: 200 }, 
      { title: 'Price', dataIndex: 'price', key: 'price', width: 100 },
      { title: 'Operations', dataIndex: '', key: 'operations', render: this.renderDelete }
    ];
  }

  componentDidMount() {
    this.props.createCheckout();
  }

  addItem = () => {
    this.props.postItem(this.state.scanCode, this.props.checkoutId);
  }

  renderDelete = (_o, row, index) => {
    return (
      <Button id={`btn-void-${row.id}`} onClick={() => {}}>Void</Button>
    );
  }

  render() {
    return(
      <div>
        <h3>Welcome, Shopper!</h3>
        <h4>Items</h4>
        <Table id='table-items'
          columns={this.itemTableColumns} 
          rowKey='id'
          data={this.props.items} 
        />
        <hr />
        <Form>
          <FormGroup>
            <Button className='btn-add' onClick={this.addItem}>Add Item</Button>
            <ControlLabel>Scan code</ControlLabel>
            <FormControl className='input-scan-code'
              onChange={ event => this.setState({ scanCode: event.target.value }) } />
          </FormGroup>
        </Form>
      </div>
    );
  }
};

const mapStateToProps = ({ checkout }) => {
  return {
    checkoutId: checkout.id,
    items: checkout.items
  };
};

export default connect(mapStateToProps, { createCheckout, postItem })(CustomerExperience);
