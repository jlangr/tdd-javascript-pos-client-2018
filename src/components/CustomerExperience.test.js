import React from 'react';
import { shallow } from 'enzyme';
import CustomerExperience from './CustomerExperience';

describe('CustomerExperience', () => {
  const component = shallow(<CustomerExperience />);
  const id = 1;

  it('renders correct', () => {
    expect(component).toMatchSnapshot();
  });

  it('initializes the `state` with an empty list of items', () => {
    expect(component.state().items).toEqual([]);
  });

  describe('when clicking `scan item`', () => {
    beforeEach(() => {
      component.setState({ items: [] });
      component.find('.btn-add').simulate('click');
    })

    it('adds a new item to `state`', () => {
      expect(component.state().items).toEqual([{ id }]);
    });

    it('adds a new item to rendered list', () => {
      expect(component.find('.item-list').children().length).toEqual(1);
    })

    it('creates an Item component', () => {
      expect(component.find('Item').exists()).toBe(true);
    });
  });

  describe('and the user wants to void item', () => {
    beforeEach(() => {
      component.setState({ items: [] });
    });

    it('removes the item from `state`', () => {
      component.find('.btn-add').simulate('click');

      component.instance().voidItem(id); // UGH id correlation

      expect(component.state().items).toEqual([]);     
    });
  });
});
