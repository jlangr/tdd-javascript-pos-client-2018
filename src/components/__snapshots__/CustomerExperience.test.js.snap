// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerExperience renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CustomerExperience />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          Welcome, Shopper!
</h3>,
        <h4>
          Items
</h4>,
        <Table
          bodyStyle={Object {}}
          columns={
                    Array [
                              Object {
                                "dataIndex": "description",
                                "key": "description",
                                "title": "Description",
                                "width": 200,
                              },
                              Object {
                                "dataIndex": "price",
                                "key": "price",
                                "title": "Price",
                                "width": 100,
                              },
                              Object {
                                "dataIndex": "",
                                "key": "operations",
                                "render": [Function],
                                "title": "Operations",
                              },
                            ]
          }
          data={Array []}
          emptyText={[Function]}
          id="table-items"
          onHeaderRow={[Function]}
          onRow={[Function]}
          prefixCls="rc-table"
          rowClassName={[Function]}
          rowKey="id"
          rowRef={[Function]}
          scroll={Object {}}
          showHeader={true}
          style={Object {}}
          useFixedHeader={false}
/>,
        <hr />,
        <Form
          bsClass="form"
          componentClass="form"
          horizontal={false}
          inline={false}
>
          <FormGroup
                    bsClass="form-group"
          >
                    <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              className="btn-add"
                              disabled={false}
                              onClick={[Function]}
                    >
                              Add Item
                    </Button>
                    <ControlLabel
                              bsClass="control-label"
                              srOnly={false}
                    >
                              Scan code
                    </ControlLabel>
                    <FormControl
                              bsClass="form-control"
                              className="input-scan-code"
                              componentClass="input"
                              onChange={[Function]}
                    />
          </FormGroup>
</Form>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Welcome, Shopper!",
        },
        "ref": null,
        "rendered": "Welcome, Shopper!",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Items",
        },
        "ref": null,
        "rendered": "Items",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bodyStyle": Object {},
          "columns": Array [
            Object {
              "dataIndex": "description",
              "key": "description",
              "title": "Description",
              "width": 200,
            },
            Object {
              "dataIndex": "price",
              "key": "price",
              "title": "Price",
              "width": 100,
            },
            Object {
              "dataIndex": "",
              "key": "operations",
              "render": [Function],
              "title": "Operations",
            },
          ],
          "data": Array [],
          "emptyText": [Function],
          "id": "table-items",
          "onHeaderRow": [Function],
          "onRow": [Function],
          "prefixCls": "rc-table",
          "rowClassName": [Function],
          "rowKey": "id",
          "rowRef": [Function],
          "scroll": Object {},
          "showHeader": true,
          "style": Object {},
          "useFixedHeader": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "form",
          "children": <FormGroup
            bsClass="form-group"
>
            <Button
                        active={false}
                        block={false}
                        bsClass="btn"
                        bsStyle="default"
                        className="btn-add"
                        disabled={false}
                        onClick={[Function]}
            >
                        Add Item
            </Button>
            <ControlLabel
                        bsClass="control-label"
                        srOnly={false}
            >
                        Scan code
            </ControlLabel>
            <FormControl
                        bsClass="form-control"
                        className="input-scan-code"
                        componentClass="input"
                        onChange={[Function]}
            />
</FormGroup>,
          "componentClass": "form",
          "horizontal": false,
          "inline": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="default"
                className="btn-add"
                disabled={false}
                onClick={[Function]}
>
                Add Item
</Button>,
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
>
                Scan code
</ControlLabel>,
              <FormControl
                bsClass="form-control"
                className="input-scan-code"
                componentClass="input"
                onChange={[Function]}
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "active": false,
                "block": false,
                "bsClass": "btn",
                "bsStyle": "default",
                "children": "Add Item",
                "className": "btn-add",
                "disabled": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add Item",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Scan code",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Scan code",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "className": "input-scan-code",
                "componentClass": "input",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            Welcome, Shopper!
</h3>,
          <h4>
            Items
</h4>,
          <Table
            bodyStyle={Object {}}
            columns={
                        Array [
                                    Object {
                                      "dataIndex": "description",
                                      "key": "description",
                                      "title": "Description",
                                      "width": 200,
                                    },
                                    Object {
                                      "dataIndex": "price",
                                      "key": "price",
                                      "title": "Price",
                                      "width": 100,
                                    },
                                    Object {
                                      "dataIndex": "",
                                      "key": "operations",
                                      "render": [Function],
                                      "title": "Operations",
                                    },
                                  ]
            }
            data={Array []}
            emptyText={[Function]}
            id="table-items"
            onHeaderRow={[Function]}
            onRow={[Function]}
            prefixCls="rc-table"
            rowClassName={[Function]}
            rowKey="id"
            rowRef={[Function]}
            scroll={Object {}}
            showHeader={true}
            style={Object {}}
            useFixedHeader={false}
/>,
          <hr />,
          <Form
            bsClass="form"
            componentClass="form"
            horizontal={false}
            inline={false}
>
            <FormGroup
                        bsClass="form-group"
            >
                        <Button
                                    active={false}
                                    block={false}
                                    bsClass="btn"
                                    bsStyle="default"
                                    className="btn-add"
                                    disabled={false}
                                    onClick={[Function]}
                        >
                                    Add Item
                        </Button>
                        <ControlLabel
                                    bsClass="control-label"
                                    srOnly={false}
                        >
                                    Scan code
                        </ControlLabel>
                        <FormControl
                                    bsClass="form-control"
                                    className="input-scan-code"
                                    componentClass="input"
                                    onChange={[Function]}
                        />
            </FormGroup>
</Form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Welcome, Shopper!",
          },
          "ref": null,
          "rendered": "Welcome, Shopper!",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Items",
          },
          "ref": null,
          "rendered": "Items",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bodyStyle": Object {},
            "columns": Array [
              Object {
                "dataIndex": "description",
                "key": "description",
                "title": "Description",
                "width": 200,
              },
              Object {
                "dataIndex": "price",
                "key": "price",
                "title": "Price",
                "width": 100,
              },
              Object {
                "dataIndex": "",
                "key": "operations",
                "render": [Function],
                "title": "Operations",
              },
            ],
            "data": Array [],
            "emptyText": [Function],
            "id": "table-items",
            "onHeaderRow": [Function],
            "onRow": [Function],
            "prefixCls": "rc-table",
            "rowClassName": [Function],
            "rowKey": "id",
            "rowRef": [Function],
            "scroll": Object {},
            "showHeader": true,
            "style": Object {},
            "useFixedHeader": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form",
            "children": <FormGroup
              bsClass="form-group"
>
              <Button
                            active={false}
                            block={false}
                            bsClass="btn"
                            bsStyle="default"
                            className="btn-add"
                            disabled={false}
                            onClick={[Function]}
              >
                            Add Item
              </Button>
              <ControlLabel
                            bsClass="control-label"
                            srOnly={false}
              >
                            Scan code
              </ControlLabel>
              <FormControl
                            bsClass="form-control"
                            className="input-scan-code"
                            componentClass="input"
                            onChange={[Function]}
              />
</FormGroup>,
            "componentClass": "form",
            "horizontal": false,
            "inline": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  className="btn-add"
                  disabled={false}
                  onClick={[Function]}
>
                  Add Item
</Button>,
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
>
                  Scan code
</ControlLabel>,
                <FormControl
                  bsClass="form-control"
                  className="input-scan-code"
                  componentClass="input"
                  onChange={[Function]}
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "active": false,
                  "block": false,
                  "bsClass": "btn",
                  "bsStyle": "default",
                  "children": "Add Item",
                  "className": "btn-add",
                  "disabled": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Add Item",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Scan code",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Scan code",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "className": "input-scan-code",
                  "componentClass": "input",
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
